File: test/operators/test_sequence_expand_op.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include <iostream>
#include "../test_include.h"
#include "operators/sequence_ops/sequence_expand_op.h"

namespace paddle_mobile {

int TestSequenceExpandOp(const framework::LoDTensor &input_x,
                         const framework::LoDTensor &input_y, int ref_level,
                         framework::LoDTensor *output) {
  VariableNameMap inputs;
  VariableNameMap outputs;
  auto scope = std::make_shared<framework::Scope>();
  inputs["X"] = std::vector<std::string>({"input_x"});
  inputs["Y"] = std::vector<std::string>({"input_y"});
  outputs["Out"] = std::vector<std::string>({"output"});

  auto input_x_var = scope.get()->Var("input_x");
  auto *x = input_x_var->template GetMutable<framework::LoDTensor>();
  x->Resize(input_x.dims());
  x->ShareDataWith(input_x);
  x->set_lod(input_x.lod());
  auto input_y_var = scope.get()->Var("input_y");
  auto *y = input_y_var->template GetMutable<framework::LoDTensor>();
  y->Resize(framework::make_ddim({0}));
  y->mutable_data<float>();
  y->set_lod(input_y.lod());

  auto output_var = scope.get()->Var("output");

  framework::AttributeMap attrs;
  attrs["ref_level"].Set<int>(0);

  auto *op = new operators::SequenceExpandOp<CPU, float>(
      "sequence_expand", inputs, outputs, attrs, scope.get());

  op->InferShape();
  op->Init();
  op->Run();

  auto *out = output_var->template Get<framework::LoDTensor>();
  output->Resize(out->dims());
  output->ShareDataWith(*out);
  output->set_lod(out->lod());
  delete op;
  return 0;
}

}  // namespace paddle_mobile

// namespace framework = paddle_mobile::framework;

int main(int argc, char *argv[]) {
  framework::LoDTensor input_x, input_y, output;
  // case 1
  {
    std::vector<float> data{1, 2, 3, 4};
    input_x.Resize(framework::make_ddim({4, 1}));
    float *in_data = input_x.mutable_data<float>();
    for (int i = 0; i < 4; ++i) in_data[i] = data[i];
    input_x.set_lod({{0, 2, 4}});
    input_y.set_lod({{0, 2, 4}, {0, 3, 6, 7, 8}});

    TestSequenceExpandOp(input_x, input_y, 0, &output);
    std::vector<float> expect_data{1, 2, 1, 2, 3, 4, 3, 4};
    std::vector<int> expect_lod{0, 2, 4, 6, 8};
    for (int i = 0; i < 5; ++i) {
      if (output.lod()[0][i] != expect_lod[i]) {
        std::cerr << "output_lod[" << i << "]: " << output.lod()[0][i]
                  << " != expect_lod[" << i << "]: " << expect_lod[i]
                  << std::endl;
        return 1;
      }
    }
    for (int i = 0; i < 8; ++i) {
      if (output.data<float>()[i] != expect_data[i]) {
        std::cerr << "output[" << i << "]: " << output.data<float>()[i]
                  << " != expect[" << i << "]: " << expect_data[i] << std::endl;
        return 1;
      }
    }
  }
  return 0;
}

File: lite/tests/unittest_py/op/test_sequence_expand_op.py
# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
sys.path.append('../')

from auto_scan_test import AutoScanTest, IgnoreReasons
from program_config import TensorConfig, ProgramConfig, OpConfig, CxxConfig, TargetType, PrecisionType, DataLayoutType, Place
import unittest
from functools import partial
import numpy as np
import hypothesis
from hypothesis import given, settings, seed, example, assume
import hypothesis.strategies as st
import argparse


class TestSequenceExpandOp(AutoScanTest):
    def __init__(self, *args, **kwargs):
        AutoScanTest.__init__(self, *args, **kwargs)
        host_places = [
            Place(TargetType.Host, PrecisionType.FP32, DataLayoutType.NCHW)
        ]
        self.enable_testing_on_place(places=host_places, thread=[1, 4])

    def is_program_valid(self,
                         program_config: ProgramConfig,
                         predictor_config: CxxConfig) -> bool:
        return True

    def sample_program_configs(self, draw):
        def generate_input(*args, **kwargs):
            if kwargs["type"] == "int32":
                return np.random.randint(kwargs["low"], kwargs["high"],
                                         kwargs["shape"]).astype(np.int32)
            elif kwargs["type"] == "int64":
                return np.random.randint(kwargs["low"], kwargs["high"],
                                         kwargs["shape"]).astype(np.int64)
            elif kwargs["type"] == "float32":
                return (kwargs["high"] - kwargs["low"]) * np.random.random(
                    kwargs["shape"]).astype(np.float32) + kwargs["low"]

        def generate_lod(seq_num, max_len):
            seq_offset = []
            sum = 0
            seq_offset.append(sum)
            for i in range(seq_num):
                sum += np.random.randint(0, max_len) + 1
                seq_offset.append(sum)
            return [seq_offset]

        input_type = draw(st.sampled_from(["int32", "int64", "float32"]))
        max_len = draw(st.integers(min_value=1, max_value=5))
        ref_level_data = draw(st.integers(min_value=-1, max_value=4))
        seq_num = draw(st.integers(min_value=1, max_value=7))
        in_shape = draw(
            st.lists(
                st.integers(
                    min_value=1, max_value=10), min_size=2, max_size=8))
        lod_x = generate_lod(seq_num, max_len)
        lod_y = generate_lod(seq_num, max_len)
        in_shape[0] = lod_x[0][-1]

        x_lod_len = lod_x[0][-1]
        y_lod_len = lod_y[0][-1]
        assume(lod_x[0][0] == lod_y[0][0])
        # x lod level is <= 1; y lod level is > 0
        assume((np.array(lod_x)).shape[0] <= 1)
        assume((np.array(lod_y)).shape[0] > 0)
        assume(ref_level_data == -1 or
               (ref_level_data >= 0 and
                ref_level_data < (np.array(lod_y)).shape[0]))
        assume((np.array(lod_x)).shape[0] == 1 and
               x_lod_len == (len(lod_y[ref_level_data])))

        sequence_expand_op = OpConfig(
            type="sequence_expand",
            inputs={"X": ["x_data"],
                    "Y": ["y_data"]},
            outputs={"Out": ["output_data"]},
            attrs={"ref_level": ref_level_data})

        program_config = ProgramConfig(
            ops=[sequence_expand_op],
            weights={},
            inputs={
                "x_data": TensorConfig(
                    data_gen=partial(
                        generate_input,
                        type=input_type,
                        low=-10,
                        high=10,
                        shape=in_shape),
                    lod=lod_x),
                "y_data": TensorConfig(
                    data_gen=partial(
                        generate_input,
                        type=input_type,
                        low=-10,
                        high=10,
                        shape=in_shape),
                    lod=lod_y)
            },
            outputs=["output_data"])

        return program_config

    def sample_predictor_configs(self):
        return self.get_predictor_configs(), ["sequence_expand"], (1e-5, 1e-5)

    def add_ignore_pass_case(self):
        pass

    def test(self, *args, **kwargs):
        self.run_and_statis(quant=False, max_examples=25)


if __name__ == "__main__":
    unittest.main(argv=[''])
